src/lib/increment_numeric.js

```diff
@@ -0,0 +1,37 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+module.exports = function incrementNumeric(x, delta) {
+    if(!delta) return x;
+
+    // Note 1:
+    // 0.3 != 0.1 + 0.2 == 0.30000000000000004
+    // but 0.3 == (10 * 0.1 + 10 * 0.2) / 10
+    // Attempt to use integer steps to increment
+    var magic = 1 / Math.abs(delta);
+    var newX = (
+        magic * x +
+        magic * delta
+    ) / magic;
+
+    // Note 2:
+    // now we may also consider rounding to cover few more edge cases
+    // e.g. 0.3 * 3 = 0.8999999999999999
+    var lenDt = ('' + delta).length;
+    var lenX0 = ('' + x).length;
+    var lenX1 = ('' + newX).length;
+
+    if(lenX1 > lenX0 + lenDt) { // likely a rounding error!
+        newX = +parseFloat(newX).toPrecision(12);
+    }
+
+    return newX;
```

Comment:
Thanks for adding those tests @archmoj - I understand (a little bit) more what you're attempting. 

----

That said, I feel like we should be extending `numFormat`:

https://github.com/plotly/plotly.js/blob/96fe2624a0df16b50ceb3061ff18d95b6a2120da/src/plots/cartesian/axes.js#L1311-L1408

where instead of exiting early

https://github.com/plotly/plotly.js/blob/96fe2624a0df16b50ceb3061ff18d95b6a2120da/src/plots/cartesian/axes.js#L1338

for when `tickformat` is set, we could construct the formatted number ourselves for `tickformat` values that include `s` and `p`. To me, that sounds a lot more robust than these floating-point tricks that gets us a rounded-off tick step. What do you think?

Reply:
I <i>think</i> it would be interesting to fix the issue at the root cause & thanks to `JavaScript`. 
These rounding issues happen in other languages too. 
Following images illustrate examples in `Fortran` & `C++`.
![fortran](https://user-images.githubusercontent.com/33888540/61755634-7f088b00-ad86-11e9-8aa7-2c5838fe9b73.png)

![cpp](https://user-images.githubusercontent.com/33888540/61755640-8334a880-ad86-11e9-9952-38cb23ba59cc.png)

Reply:
> it would be interesting to fix the issue at the root cause

What is the _root_ cause in your mind here?